### PR TITLE
Jackie's MediaRanker

### PR DESCRIPTION
**Pull request by jfahmy**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. |  I wrote a class method to retrieve the work with the most votes (self.top_work). It gathers all the works with Work.all and then uses a max_by to find the work with the most votes. It returns this Work to the Works controller so that it can be displayed at the top of my hope page as the featured work. 
Describe how you approached testing that model method. What edge cases did you come up with? | For testing the works model methods I tested that is would return an empty collection or a collection with a length of 0 if it was given a category for which no works exist. I also tested that a work without any votes will still be returned by my work model methods. I could probably use more. My vote validation tests are probably stronger in terms of edge cases. 
What are `session` and `flash`? What is the difference between them? | Session stores cookies that can be accessed by the browser as a user moves from page to page. Flash stores messages temporarily in a flash hash that can be served to the user immediately after they've executed an action on a page.  
Describe a controller filter you wrote. |  I used a controller filter method in my user controller, in my work controller and in my application controller. In all three cases it is a private method that is being run ahead of the other specified controller methods.  (get_user, get_work, and find_user) These methods dry out the code so that I do not need to do a .find or .find_by inside individual controller methods, the instance variables I need for this objects are automagically available to me.
What was one thing that you gained more clarity on through this assignment? |  Routes, how to create model methods, and validations. I used scope as part of my validations to ensure that a user couldn't vote on the same work twice and that seems like it will be useful in the future.  I also understand how to use inspect in the browser much better now from all the css copy and pasting :)
What is the Heroku URL of your deployed application? | https://jackies-media-ranker.herokuapp.com/works
Do you have any recommendations on how we could improve this project for the next cohort? |  Maybe the instruction to try to make your website look as much like the real MediaRanker as possible is not the best for learning bootstrap. I understand how to make a bootstrap button, (btn btn-primary!) but that's really the only css I've internalized.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Rails.application.routes.draw do
+  root 'works#main'
+
+  post 'sessions/login', to: 'sessions#login', as: 'login'
+  get 'sessions/login', to: 'sessions#new'
+  delete 'sessions/destroy', to: 'sessions#destroy', as: 'logout'
+
+  get 'works/main', to: 'works#main'
```
Comment (CheezItMan) [inline config/routes.rb:8]: Since you have `sessions#main` being routed to by the root_path, you don't really need this route.

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+  <table class="table">
+    <tr>
+      <th> Votes </th>
+      <th> Title </th>
+      <th> Created By </th>
+      <th> Published </th>
+      <th> Upvote </th>
+    </tr>
+  <% work.each do |work| %>
```
Comment (CheezItMan) [inline app/views/works/index.html.erb:15]: Notice how these tables are pretty repetitive.  You can DRY them out by using partials.

File: test/models/user_test.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    it "is invalid if the username is not unique" do
+      user1 = users(:user1)
+      user2 = User.create(username: "Vinnie")
+
+      value(user1.valid?).must_equal true
+      value(user2.valid?).must_equal false
+    end
+
+    it "is invalid if the username is not between 5 and 20 characters" do
```
Comment (CheezItMan) [inline test/models/user_test.rb:25]: You should test the username at 4 letters and then 5, verifying that one is invalid and the other valid.  Then verify that 20 characters are valid and 21 invalid.  

Check the edge cases.

File: test/models/vote_test.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    it 'will not create a vote without a user' do
+      work = works(:album)
+      vote.work_id = work.id
+      vote.save
+
+      expect(vote.errors.messages).must_include :user
+    end
+
+    it 'allows user to create a vote' do
```
Comment (CheezItMan) [inline test/models/vote_test.rb:20]: This should be named `it 'is valid with both a user and a work'`

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        session[:user_id] = @user.id
+        flash[:success] = "Pleased to meet you #{@user.username}! Start voting!"
+        redirect_to root_path
+      else
+        flash.now[:warning] = "No username entered. Try again."
+        render :new
+      end
+    else
+      session[:user_id] = @user.id
+      flash[:success] = "Welcome back #{user.username}!"
```
Comment (CheezItMan) [inline app/controllers/sessions_controller.rb:17]: Bug here, this should be `@user.username`

**Comment by CheezItMan** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | More commits would be good, but you have good commit messages
Comprehension questions	|  Check, 
**General** |
Rails fundamentals (RESTful routing, use of named paths) |  Really good here
Views are well-organized (DRY, use of semantic HTML, use of partials) | Check, but you could DRY them out further with some partial views.  
Errors are reported to the user |  Check
Business logic lives in the models |  Check, nice work
Models are thoroughly tested, including relations, validations and any custom logic |  Check, well done
**Wave 1 - Media** |
Splash page shows the three media categories |  Check
Basic CRUD operations on media are present and functional |   Check
**Wave 2 - Users and Votes** |
Users can log in and log out |  **Bug** Yes, but I can only log in and out once, logging in the 2nd time tips a syntax error.  
The ID of the current user is stored in the session | Check
A user cannot vote for the same media more than once | Check
All media lists are ordered by vote count | Check
Splash page contains a media spotlight | Check
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | Check
Individual user pages and the user list are present |  Check
**Optional - Styling** |
Bootstrap is used appropriately |  Well Done
Look and feel is similar to the original |  Looks much like the original
**Overall** |  Excellent work, you hit all the learning goals.  There are minor issues, but you hit all the highlights.  See my comments and let me know any questions.
